src/authenticators/mod.rs

```diff
@@ -46,11 +49,65 @@ impl Deref for ApplicationName {
     }
 }
 
+#[allow(deprecated)]
+#[deprecated(since = "0.9.0", note = "ApplicationIdentity should be used instead.")]
+impl ApplicationName {
+    /// Create ApplicationName from name string only
+    pub fn from_name(name: String) -> ApplicationName {
+        ApplicationName { name }
+    }
+}
+
+#[allow(deprecated)]
+impl std::fmt::Display for ApplicationName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.name)
+    }
+}
+
+#[allow(deprecated)]
+impl From<Application> for ApplicationName {
+    fn from(app: Application) -> Self {
+        ApplicationName::from_name(app.identity.name)
+    }
+}
+
+/// A unique identifier for an application.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ApplicationIdentity {
+    /// The name of the application.
+    pub name: String,
+    /// The id of the authenticator used to authenticate the application name.
+    pub authenticator_id: u8,
```

Comment:
Maybe it'd be possible to use `Option<AuthType>` here, or some wrapper type around `AuthType` so that we don't just have the special case values (like `0xFF`) floating around. Maybe an enum, something like
```rust
enum AuthenticatorIdentity {
    KnownType(AuthType),
    UnknownType,
}
```
@hug-dev - thoughts?

Reply:
We would need something like `AuthType::Unknown` adding to the parsec interface.

Reply:
```
pub enum AuthType {
    /// No authentication
    NoAuth = 0,
    /// Direct authentication
    Direct = 1,
    /// JSON Web Tokens (JWT) authentication (not currently supported)
    Jwt = 2,
    /// Unix peer credentials authentication
    UnixPeerCredentials = 3,
    /// Authentication verifying a JWT SPIFFE Verifiable Identity Document
    JwtSvid = 4,
    /// The authentication of an incoming request is unknown.
    IngressAuthUnknown = 0xFE
    /// The authentication of a stored key is unknown.
    EgressAuthUnknown = 0xFF
}
```
Maybe something like this? Just a suggestion

Reply:
Hmmm, I'd keep the "unknown" stuff out of `AuthType`, I think Hugues was asking why we can't use `AuthType` as it is now. We can't because there are places (e.g. when reading from disk in the on-disk KIM) where we don't know what authentication was used, but we need to populate with something. The initial option was to just use `NoAuth`, but I don't think that's sound because you could end up actually confusing it with `NoAuth` as an actual authentication "mechanism" somewhere.

Reply:
Suppose in the future we update `list_keys` and the `KeyInfo` object to return our new data types. Would it not be better to have a "we don't know what authentication this key was created with" type in the interface? I do agree we should have something different to `NoAuth`

Reply:
After further investigation, `IngressAuthUnknown` isn't really needed as I can just update the functions where authentication information is currently missing.
```
pub enum AuthType {
    /// No authentication
    NoAuth = 0,
    /// Direct authentication
    Direct = 1,
    /// JSON Web Tokens (JWT) authentication (not currently supported)
    Jwt = 2,
    /// Unix peer credentials authentication
    UnixPeerCredentials = 3,
    /// Authentication verifying a JWT SPIFFE Verifiable Identity Document
    JwtSvid = 4,
    /// The authentication type of a stored key is unknown.
    /// In older versions of Parsec the authentication method was not stored.
    /// Therefore, a key with AuthType::Unknown *could* have been created with
    /// authentication or with NoAuth; neither state can however be guaranteed
    /// therefore AuthType::Unknown will be returned to avoid collisions.
    Unknown = 0xFF
}
```
Something like this?